baram/ec2_manager.py

```diff
@@ -1,105 +1,306 @@
+
+        result = []
+        for i in range(len(sg_rules)):
+            result.append({'security_group_id': sg_ids[i],
+                           'is_ob': is_obs[i],
+                           'related_security_group_id': referenced_groups[i]})
+
+        return result
+
+    def get_related_security_groups(self, security_group_id: str):
+        """
+        Get the set of related security groups for specific security group.
+
+        :param security_group_id: GroupId
+        :return: GroupId
+        """
+        sg_relations = self.list_security_group_relations()
+
+        result = [sg_relation['security_group_id'] for sg_relation in sg_relations
+                  if sg_relation['related_security_group_id'] == security_group_id]
+        return set(result)
+
+    def get_security_group_rules(self, security_group_id: str):
+        """
+        Get security group rules of specific security group.
+
+        :param security_group_id: GroupId
+        :return: SecurityGroupRuleId
+        """
+        sg_rules = self.cli.describe_security_group_rules()['SecurityGroupRules']
+
+        result = [{'security_group_rule_id': sg_rule['SecurityGroupRuleId'],
+                   'is_ob': True if sg_rule['IsEgress'] else False}
+                  for sg_rule in sg_rules if sg_rule['GroupId'] == security_group_id]
+        return result
+
+    def revoke_security_group_rules(self, security_group_id_list: list):
+        """
+        Get rid of security group rule of security groups in list.
+
+        :param security_group_id_list: GroupId
+        """
+        for sg_id in security_group_id_list:
+            self.revoke_security_group_rule(sg_id)
+
+    def revoke_security_group_rule(self, security_group_id: str):
+        """
+        Get rid of security group rule of specific security group.
+
+        :param security_group_id: GroupId
+        """
+        sg_rules = self.get_security_group_rules(security_group_id)
+
+        try:
+            for sg_rule in sg_rules:
+                if sg_rule['is_ob']:
+                    self.cli.revoke_security_group_egress(GroupId=security_group_id,
+                                                          SecurityGroupRuleIds=[sg_rule['security_group_rule_id']])
+                else:
+                    self.cli.revoke_security_group_ingress(GroupId=security_group_id,
+                                                           SecurityGroupRuleIds=[sg_rule['security_group_rule_id']])
+                print(f"{security_group_id}'s {list(sg_rule.values())[0]} rule is deleted")
+        except:
+            self.logger.info('error')
+
+    def delete_security_groups(self, security_group_id_list: list):
+        """
+        Delete every security groups in input list.
+
+        :param security_group_id_list: GroupId
+        """
+        for sg_id in security_group_id_list:
+            self.delete_security_group(sg_id)
+
+    def delete_security_group(self, security_group_id: str):
```

Comment:
Is there a certain reason for separating `delete_security_groups` and `delete_security group`?
Only leaving `delete_security_groups` and merging `delete_security group`'s functions(try-except, logger) could be an option.

Reply:
It was kinda trace of making `delete_redundant_security_groups`. I'll actively consider your comment

Reply:
Maybe `delete_redundant_security_groups`, `delete_security_groups`, `delete_security_group` could be integrated because `list_redundant_security_groups` already exists.